independent-projects/tools/codestarts/codestarts.adoc

```diff
@@ -0,0 +1,220 @@
+
+ Example: in the Quarkus project `codestart.yml`:
+[source,yaml]
+----
+output-strategy:
+  "pom.xml": smart-pom-merge # smartly merge pom.xml contents together
+  "README.md": append # they are appended together
+  "readme.md": forbidden # it will fail if found in any codestart
+  "README.adoc": forbidden
+  "readme.adoc": forbidden
+  ".gitignore": append
+  "src/main/resources/application.yml": smart-config-merge # smartly merge config together and automatically converted to the selected config type (yaml or properties)
+  "src/main/resources/application.yaml": forbidden # should use yml instead
+  "src/main/resources/application.properties": forbidden # should use yml instead
+  "*": fail-on-duplicate # it will fail if a duplicate is found for all other files
+----
+
+* other files are just copied as is.
+
+
+=== Writing Example Code for an Extension
+
+Let's imagine we are writing the example codestart for my "foo" extension. As we discussed before, the base of the project is already provided so we can focus on the example code.
+
+1. Create your `codestart.yml`:
+
+codestart.yml:
+[source,yaml]
+----
+---
+name: foo-example
+ref: foo
+type: example
+missing-languages:
+  - kotlin
+  - scala
+spec:
+  base:
+    data:
+      # Some data for my templates which gives could be overridden by the user input
+      some:
+        data:
+          here: bar
+          there: foo
+    dependencies:
+      # the dependencies needed for my codestart (note that the extension dependency is auto-added if missing)
+      - io.quarkus:quarkus-foo
+      - io.quarkus:quarkus-resteasy
+      # ... you can also provide a version (when it's not part of the platform bom)
+      - group:artifact:version
+    test-dependencies:
+      # test dependencies
+      - io.rest-assured:rest-assured
+----
+
+*NOTE* You don't need to add `pom.xml` or `build.gradle`, it is going to be auto generated. It will include the dependencies provided in the `codestart.yml`.
+
+2. Add you example code in java, kotlin or scala
+
+* java/src/main/java/org/acme/foo/Foo.java:
+[source,java]
+----
+package org.acme.foo
+
+import javax.ws.rs.GET
+import javax.ws.rs.Path
+import javax.ws.rs.Produces
+import javax.ws.rs.core.MediaType
+
+@Path("foo")
+class ExampleResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    fun hello() = "Hello foo"
+}
+----
+
+* kotlin/src/main/kotlin/org/acme/foo/Foo.kt
+* scala/src/main/scala/org/acme/foo/Foo.kt
+
+*NOTE* Just add `.tpl.qute` if you want it to be rendered with qute and use some data
+
+*NOTE* Most of the time you won't need it for code, but you can use a `base` directory to add files to process for all languages.
+
+
+2. Add some configuration (if needed)
+
+The `base/src/main/resources/application.yml` file is going to be merged with the other codestarts config and automatically converted to the selected config type (yaml or properties).
```

Comment:
what happens if two extensions list the same key ? warning, error, nothing ?

Reply:
For now, it's silently overridden depending on extension ordering which is close to gravity :) I guess we should fail?

Reply:
warning just to not make it so a rogue extension blocks everything.

Reply:
I'll do that in another PR

Reply:
https://github.com/quarkusio/quarkus/issues/11083